src/topology/alexandroff.lean

```diff
@@ -0,0 +1,322 @@
+  is_open (of⁻¹' s) :=
+begin
+  by_cases H : ∞ ∈ s,
+  { simpa using ((is_open_iff_of_mem H).mp h).2 },
+  { exact (is_open_iff_of_not_mem H).mp h }
+end
+
+end topology
+
+section topological_prop
+
+variables {X : Type*} [topological_space X]
+
+@[continuity] lemma continuous_of : continuous (@of X _) :=
+continuous_def.mpr (λ s hs, is_open_of_is_open hs)
+
+/-- An open set in `alexandroff X` constructed from a closed compact set in `X` -/
+def opens_of_compl {s : set X} (h : is_compact s ∧ is_closed s) :
+  topological_space.opens (alexandroff X) :=
+⟨(of '' s)ᶜ, by { rw [is_open_iff_of_mem ((mem_compl_iff _ _).mpr infty_not_mem_image_of),
+  preimage_compl, compl_compl, of_injective.preimage_image _], exact h }⟩
+
+lemma infty_mem_opens_of_compl {s : set X} (h : is_compact s ∧ is_closed s) :
+  ∞ ∈ (opens_of_compl h) :=
+by { simp only [opens_of_compl, topological_space.opens.coe_mk],
+     exact mem_compl infty_not_mem_image_of }
+
+lemma is_open_map_of : is_open_map (@of X _) :=
+λ s hs, begin
+  rw [← preimage_image_eq s of_injective] at hs,
+  rwa is_open_iff_of_not_mem infty_not_mem_image_of
+end
+
+lemma is_open_range_of : is_open (@range_of X _) :=
+is_open_map_of.is_open_range
+
+instance : compact_space (alexandroff X) :=
+{ compact_univ :=
+  begin
+    refine is_compact_of_finite_subcover (λ ι Z h H, _),
+    simp only [univ_subset_iff] at H ⊢,
+    rcases Union_eq_univ_iff.mp H ∞ with ⟨K, hK⟩,
+    have minor₁ : is_compact (of⁻¹' Z K)ᶜ,
+    { specialize h K, rw is_open_iff_of_mem hK at h, exact h.1 },
+    let p : ι → set X := λ i, of⁻¹' Z i,
+    have minor₂ : ∀ i, is_open (p i) := λ i, is_open_of_is_open (h i),
+    have minor₃ : (of⁻¹' Z K)ᶜ ⊆ ⋃ i, p i :=
+      by simp only [p, ← preimage_Union, H, preimage_univ, subset_univ],
+    rcases is_compact_iff_finite_subcover.mp minor₁ p minor₂ minor₃ with ⟨ι', H'⟩,
+    refine ⟨insert K ι', _⟩,
+    rw ← preimage_compl at H',
+    simp only [Union_eq_univ_iff],
+    intros x,
+    by_cases hx : x ∈ Z K,
+    { exact ⟨K, mem_Union.mpr ⟨finset.mem_insert_self _ _, hx⟩⟩ },
+    { have triv₁ : x ≠ ∞ := (ne_of_mem_of_not_mem hK hx).symm,
+      rcases ne_infty_iff_exists.mp triv₁ with ⟨y, hy⟩,
+      have triv₂ : (y : alexandroff X) ∈ {x} := mem_singleton_of_eq hy.symm,
+      rw [← mem_compl_iff, ← singleton_subset_iff] at hx,
+      have : of⁻¹' {x} ⊆ of⁻¹' (Z K)ᶜ := λ y hy, hx hy,
+      have key : y ∈ ⋃ (i : ι) (H : i ∈ ι'), p i := this.trans H' (mem_preimage.mpr triv₂),
+      rcases mem_bUnion_iff'.mp key with ⟨i, hi, hyi⟩,
+      refine ⟨i, mem_Union.mpr ⟨finset.subset_insert _ ι' hi, _⟩⟩,
+      simpa [hy] using hyi }
+  end }
+
+lemma dense_range_of (h : ¬ is_compact (univ : set X)) : dense (@range_of X _) :=
+begin
+  refine dense_iff_inter_open.mpr (λ s hs Hs, _),
+  by_cases H : ∞ ∈ s,
+  { rw is_open_iff_of_mem H at hs,
+    have minor₁ : s ≠ {∞},
+    { by_contra w,
+      rw [not_not.mp w, of_preimage_infty, compl_empty] at hs,
```

Comment:
`by_contra` followed by `rw not_not` is just `intro`:
```suggestion
    { intro hw,
      rw [hw, of_preimage_infty, compl_empty] at hs,
```

Reply:
Or even
```suggestion
    { rintro rfl,
      rw [of_preimage_infty, compl_empty] at hs,
```

Reply:
Thank you for your reviews! I could learn a lot from them. What does `rintro rfl` do? I guess it's like saying intro some equality and register it as a `rfl`?

Reply:
`rintro rfl` is like doing a pattern match on `| rfl :=` in the equation compiler. You can read it as roughly `intro w`, `subst w`.